pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/InMemoryDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,184 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.delayed;
+
+import io.netty.util.Timeout;
+import io.netty.util.Timer;
+import io.netty.util.TimerTask;
+
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers;
+import org.apache.pulsar.common.util.collections.TripleLongPriorityQueue;
+
+@Slf4j
+public class InMemoryDelayedDeliveryTracker implements DelayedDeliveryTracker, TimerTask {
+
+    private final TripleLongPriorityQueue priorityQueue = new TripleLongPriorityQueue();
+
+    private final PersistentDispatcherMultipleConsumers dispatcher;
+
+    // Reference to the shared (per-broker) timer for delayed delivery
+    private final Timer timer;
+
+    // Current timeout or null if not set
+    private Timeout timeout;
+
+    // Timestamp at which the timeout is currently set
+    private long currentTimeoutTarget;
+
+    private final long tickTimeMillis;
+
+    InMemoryDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispatcher, Timer timer, long tickTimeMillis) {
+        this.dispatcher = dispatcher;
+        this.timer = timer;
+        this.tickTimeMillis = tickTimeMillis;
+    }
+
+    @Override
+    public boolean addMessage(long ledgerId, long entryId, long deliveryAt) {
+        long now = System.currentTimeMillis();
```

Comment:
it is better to use jdk `Clock` here. so the constructor takes a parameter `Clock` and it calls `clock.millis()` to get the current timestamp. The reason of using `Clock` is that it is much easier to test.

Reply:
Adding the Clock and mocked unit tests for the tracker in isolation.